miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/file/MiniAppFileDownloader.kt

```diff
@@ -0,0 +1,146 @@
+
+    fun setBridgeExecutor(activity: Activity, bridgeExecutor: MiniAppBridgeExecutor) {
+        this.activity = activity
+        this.bridgeExecutor = bridgeExecutor
+    }
+
+    private fun <T> whenReady(callback: () -> T) {
+        if (this::bridgeExecutor.isInitialized && this::activity.isInitialized) {
+            callback.invoke()
+        }
+    }
+
+    internal fun onStartFileDownload(callbackObj: CustomFileDownloadCallbackObj) = whenReady {
+        val fileName = callbackObj.param?.filename ?: ""
+        val url = callbackObj.param?.url ?: ""
+        val headers = callbackObj.param?.headers
+        val successCallback =
+            { fileName: String -> bridgeExecutor.postValue(callbackObj.id, fileName) }
+        val errorCallback = { message: String ->
+            bridgeExecutor.postError(callbackObj.id, "$ERR_FILE_DOWNLOAD $message")
+        }
+        scope.launch {
+            startDownloading(fileName, url, headers, successCallback, errorCallback)
+        }
+    }
+
+    private fun startDownloading(
+        fileName: String,
+        url: String,
+        headers: DownloadFileHeaderObj?,
+        successCallback: (String) -> Unit,
+        errorCallback: (String) -> Unit
+    ) {
+        val file = createFileDirectory(fileName = fileName)
+        val extension = url.substring(url.lastIndexOf("."))
+        val mimetype = mimeTypeMap.getMimeTypeFromExtension(extension)
+        val client = createHttpClient(headers)
+        val request = Request.Builder().url(url).build()
+        val response = client.newCall(request).execute()
+        if (response.isSuccessful) {
+            response.body?.let { responseBody ->
+                writeInputStreamToFile(inputStream = responseBody.byteStream(), file = file)
+                openShareIntent(mimetype, file)
+                successCallback(fileName)
+            } ?: run {
+                errorCallback(ERR_EMPTY_RESPONSE_BODY)
+            }
+        } else {
+            errorCallback("Error Code ${response.code}")
+        }
+    }
+
+    private fun createHttpClient(headers: DownloadFileHeaderObj?): OkHttpClient {
+        val builder = OkHttpClient.Builder()
+        headers?.token?.let { token ->
+            builder.addNetworkInterceptor {
+                val requestBuilder = it.request().newBuilder()
+                requestBuilder.addHeader("token", token)
+                it.proceed(requestBuilder.build())
+            }
+        }
+        return builder.build()
+    }
+
+    private fun createFileDirectory(fileName: String): File {
+        val cacheDir = File("${activity.cacheDir}/mini_app_download")
+        if (!cacheDir.exists()) {
+            cacheDir.mkdirs()
+        }
+        cache = cacheDir
+        return File(cacheDir, fileName)
+    }
+
+    @Suppress(" NestedBlockDepth", "MagicNumber")
+    private fun writeInputStreamToFile(inputStream: InputStream, file: File) {
+        inputStream.use { inputStream ->
```

Comment:
We can write `inputStream.use { input ->` to avoid shadowness.

<img width="287" alt="Screen Shot 2022-03-17 at 21 21 44" src="https://user-images.githubusercontent.com/66930373/158807580-48ace79a-6ca1-4564-8303-8c8082be8a15.png">